### ISSUE TITLE
[no-unnecessary-type-assertion] False positives/conflict with typescript

### ISSUE THREAD

**Issue by whalderman** [typescript-eslint/typescript-eslint#2174]
**Repro**

Default ruleset.

```TS
const document = new Document(); // for the Document type
const styles = Array.from(document.styleSheets) as CSSStyleSheet[]; // eslint(@typescript-eslint/no-unnecessary-type-assertion)
styles.map((stylesheet: CSSStyleSheet) => { // if the assertion above is removed, this reports a ts(2345) error (argument of type '...' is not assignable to parameter of type '???')
	const rules = Array.from(stylesheet.rules); // If the type guide is removed from the parameter above, "Property 'rules' does not exist on type 'StyleSheet'"
});
```

**Expected Result**
I assume it should at least be allowed, the assertion that each stylesheet is a CSSStyleSheet is required for the following map()....

**Actual Result**
If asserted, eslint error.
If removed, typescript error.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `14.4.0` |
| `npm`                              | `6.13.7` |

Perhaps I'm missing something, if anyone could confirm the issue or point out an error I made, I'd appreciate it!

**Comment by essenmitsosse** [comment]
I think all of this is intendet behaviour. The rule is warning you that you shouldn't unnecessarily typecast. And `StyleSheet` not automatically being `CSSStyleSheet` is also intended on TypeScript's side:
https://github.com/microsoft/TypeScript/issues/15319

The correct way as far as I can see would be to add a type guard, somehow like this:
```TS
Array
    .from(new Document().styleSheets)
    .map((stylesheet) => {
        if (stylesheet instanceof CSSStyleSheet){
            return Array.from(stylesheet.rules);
        }
        throw new Error("This shouldn't happen or what ever ...")
    });
```

**Comment by whalderman** [comment]
Ah, that makes sense. Probably would have taken days for me to find... Thanks for pointing that out for me!
